spec/models/concerns/message_subscriptions_spec.rb

```diff
@@ -0,0 +1,214 @@
+          let!(:supervisor) { create :user, :supervisor }
+          before { User.current_user = supervisor }
+
+          it 'should subscribe a supervisor for subsequent messages' do
+            # The unrelated supervisor receives the first message of the thread
+            expect(message1).to receive(:add_subscription).with('read', reviewer.id)
+            expect(message1).to receive(:add_subscription).with('unread', supervisor.id)
+            message1.subscribe_users_to_message
+
+            # The supervisor answers on the the private thread
+            create :message, sender: supervisor, messageable: offer, is_private: true
+
+            # The supervisor receives subsequent message of the thread
+            expect(message2).to receive(:add_subscription).with('read', reviewer.id)
+            expect(message2).to receive(:add_subscription).with('unread', supervisor.id)
+            message2.subscribe_users_to_message
+          end
+
+          it 'should subscribe a reviewer for subsequent messages' do
+            other_reviewer = create(:user, :reviewer)
+            User.current_user = other_reviewer
+            # The unrelated supervisor receives the first message of the thread
+            expect(message1).to receive(:add_subscription).with('read', reviewer.id)
+            expect(message1).to receive(:add_subscription).with('unread', other_reviewer.id, )
+            message1.subscribe_users_to_message
+
+            # The supervisor answers on the the private thread
+            create :message, sender: other_reviewer, messageable: offer, is_private: true
+
+            # The supervisor receives subsequent message of the thread
+            expect(message2).to receive(:add_subscription).with('read', reviewer.id) # sender
+            expect(message2).to receive(:add_subscription).with('unread', other_reviewer.id)
+            message2.subscribe_users_to_message
+          end
+        end
+      end
+
+      context 'public thread' do
+        let!(:supervisor) { create :user, :supervisor }
+        before { User.current_user = supervisor }
+        it 'should subscribe a supervisor for subsequent messages' do
+          # The unrelated supervisor receives the first message of the thread
+          expect(message1).to receive(:add_subscription).with('read', reviewer.id)
+          expect(message1).to receive(:add_subscription).with('unread', supervisor.id)
+          message1.subscribe_users_to_message
+
+          # The supervisor answers on the the private thread
+          create :message, sender: supervisor, messageable: message.offer, is_private: false
+
+          # The supervisor receives subsequent message of the thread
+          expect(message2).to receive(:add_subscription).with('read', reviewer.id)
+          expect(message2).to receive(:add_subscription).with('unread', supervisor.id)
+          message2.subscribe_users_to_message
+        end
+      end
+    end
+
+    describe '#handle_subscriptions' do
```

Comment:
this method doesn't exist any more. I think you can move the spec below for `should subscribe mentioned_users` into the `private messages` block above.